README.md
# Casim.ir

## Yet Another URL Shortener (or Shrinker)

Casim.ir is inspired by — and tries to find its way between — both [lilURL](http://lilurl.sourceforge.net/) and [xav.cc](http://xav.cc/) 's [sfShortUrlPlugin](http://www.symfony-project.org/plugins/sfShortUrlPlugin), the first lacking some essential features and the later needing a whole [symfony](http://www.symfony-project.org/) environment, as the location implies.

Casim.ir is both available online at [casim.ir](http://casim.ir/) and as a [download](https://github.com/chusopr/casimir/releases) for self hosting.

It only needs PHP and MySQL to run. Just install it, following the instructions in the readme file, and enjoy it!

The development is [hosted on GitHub](http://github.com/chusopr/casimir), so you can fork it and help me enhance it at will!

*Disclaimer:* I didn't try to follow any PHP golden development rules, it just had to work as I needed.

## License

See the included [LICENSE.md](https://github.com/chusopr/casimir/blob/master/LICENSE.md) file.

## Installation

### Database creation

Create these tables in your MySQL database:

    CREATE TABLE `casimir` (
      `short_url` varchar(50) NOT NULL DEFAULT '',
      `long_url` text NOT NULL,
      `creation_date` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
      `last_use_date` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
      `uses` int(11) unsigned NOT NULL DEFAULT '0',
      PRIMARY KEY (`short_url`)
    ) ENGINE=MyISAM DEFAULT CHARSET=utf8;
  
    CREATE TABLE `casimir_stats` (
      `short_url` varchar(50) NOT NULL default '',
      `use_date` timestamp NOT NULL default '0000-00-00 00:00:00'
    ) ENGINE=MyISAM DEFAULT CHARSET=utf8;

### Configuration

Copy `user/casimir-conf.php.example` to `user/casimir-conf.php` and edit the configuration settings.

Rename `htaccess` to `.htaccess` according to your settings.

### Personalization

Optionnaly, you can add
- a `user/footer.php` script that will be added after Casim.ir's footer
- a `user/screen.css` stylesheet that will be loaded after Casim.ir's one


changelog.markdown
# Cas.im/ir v2.x

## Cas.im/ir v2.2 (2020/10/06)

- Add support to customize site title and description.
- Disable API when CAPTCHA is enabled (fixed security flaw allowing using the API to bypass CAPTCHA).
- Fixed code indentation.
- Return 404 Not Found when a non-existent URL is requested.
- Add support for hCaptcha.
- Add support for CAPTCHA localization.
- Fix indentation inconsistencies.

## Cas.im/ir v2.1 (2014/12/04)

- New maintainer (#15)
- Use mysqli extension instead of deprecated mysql\_\* functions (#14)
- CAPTCHA support using reCAPTCHA.net
- Updated project URLs
- Set up new test site since the old one has expired (https://github.com/chusopr/casimir/issues/15#issuecomment-63443518, #17)

## Cas.im/ir v2.0 (2012/02/07)

Diet. KISS.

If you upgraded from a previous release, you can remove the `title_url` field from the `casimir` table:
    ALTER TABLE `casimir` DROP COLUMN `title_url`

- FIXED: PHP 5.3 compatibility (ereg deprecated)
  cf https://github.com/neofutur/gwgd/commit/bd9e52e9f2a983a779bd10f33c5b780d45883158
- FIXED: no longer fails miserably when there is no configuration file yet

# Cas.im/ir v1.x

## Cas.im/ir v1.3 (2011/01/17)

Enhancements and fixes by [Neofutur](https://github.com/neofutur):

If you don't have it yet, add this field to your "casimir" table:
    ALTER TABLE `casimir` ADD `title_url` VARCHAR(128) NOT NULL DEFAULT 'No title defined for this url', ADD FULLTEXT (`title_url`)

## Cas.im/ir v1.2 (2010/01/20)

- FIXED: handle empty short URL creation with bookmarklet
- FIXED: readme.txt

## Cas.im/ir v1.1 (2010/01/05)

- NEW: daily, weekly, and monthly statistics
- FIXED: access_key added to the bookmarklet

If you don't have it yet, create this table in your MySQL database:

    CREATE TABLE `casimir_stats` (
      `short_url` varchar(50) NOT NULL default '',
      `use_date` timestamp NOT NULL default '0000-00-00 00:00:00'
    ) ENGINE=MyISAM DEFAULT CHARSET=utf8



inc/Casimir.php
<?php

// Failback _() function if gettext is not supported
if (!function_exists("_"))
{
  function _($s)
  {
    return $s;
  }
}

class Casimir {
  public $version;
  public $base_url;
  public $short;
  public $msg;
  public $ok;
  public $access_key;
  private $db;
  private $locale;
  private $captcha_service;

  function __construct() {
    $this->version = '2.2';
    $this->db = new mysqli(MYSQL_HOST, MYSQL_USER, MYSQL_PASSWORD, MYSQL_DATABASE);
    if ($this->db->connect_error) die(_('Could not connect to database'));
    $current_dir = dirname($_SERVER['PHP_SELF']);
    if ($current_dir == '/') $current_dir = '';
    $this->base_url = 'http://'.$_SERVER['SERVER_NAME'].$current_dir.'/';
    $this->short = '';
    $this->msg = '';
    $this->ok = true;
    $this->access_key = '';
    if (isset($_GET['access_key'])) {
      $this->access_key = $_GET['access_key'];
    } elseif (isset($_POST['access_key'])) {
      $this->access_key = $_POST['access_key'];
    }
    $this->setLocale();
    if (defined('RECAPTCHA') && RECAPTCHA)
      $this->captcha_service = RECAPTCHA == 'hcaptcha'? 'h-':'g-re';
  }

  private function tryLocale($l)
  {
    if (empty($l)) return false;

    // First, test if we support this locale
    if (
         // We may support it with country code: xx_XX
         (!file_exists("locale/$l/LC_MESSAGES/casimir.mo")) &&
         // Or as a country-less language code: xx
         (!file_exists("locale/" . preg_replace("/_.*/", "", $l) . "/LC_MESSAGES/casimir.mo"))
       )
      return false;

    if (!@setlocale(LC_ALL, "$l.utf8"))
      return false;

    putenv("LC_ALL=$l");

    bindtextdomain("casimir", "locale");
    textdomain("casimir");

    return true;
  }

  private function setLocale()
  {
    $this->locale = false;
    // Check if user has provided its list of preferred languages
    if (empty($_SERVER["HTTP_ACCEPT_LANGUAGE"]))
      return false;

    $this->locale = Locale::acceptFromHttp($_SERVER['HTTP_ACCEPT_LANGUAGE']);
    // Let's try if we support user preferred language
    if (
         (class_exists("Locale")) &&
         ($this->tryLocale($this->locale))
       )
    return true;

    // Get list of preferred languages as provided by user to try them all
    $languages = explode(",", $_SERVER["HTTP_ACCEPT_LANGUAGE"]);
    if ((!empty($languages)) && (is_array($languages)))
      // Iterate over languages to find the first we support
      foreach ($languages as $l)
      {
        // Convert Accept-Language language to a locale one
        $l = preg_replace(array("/;.*/", "/-/"), array("", "_"), $l);
        if ($this->tryLocale($l))
          return true;
      }

    return false;
  }

  function handleRequest($api) {
    if (preg_match("#^.*/\??([^=]+)$#i", $_SERVER['REQUEST_URI'], $regs)) {
      $this->short = $this->db->real_escape_string($regs[1]);
    } else {
      $this->short = '';
    }
    if ($this->short != '' && $this->short != basename($_SERVER['PHP_SELF'])) {
      if ($location = $this->getLong($this->short)) {
        $this->updateUses($this->short);
        header('Status: 301 Moved Permanently', false, 301);
        header('Location: '.$location);
        exit;
      } else {
        header('Status: 404 Not Found', true, 404);
        $this->ok = false;
        $this->msg = _('Sorry, but this short URL isn\'t in our database.');
      }
    }

    if (defined('ACCESS_KEY') && ACCESS_KEY != '' && ACCESS_KEY != $this->access_key) {
      $this->ok = false;
      $this->msg = _('This Casimir instance is protected, you need an access key!');
    } else {
      if (isset($_POST['long'])) {
        list($this->ok, $this->short, $this->msg) = $this->addUrl($_POST['long'], isset($_POST['short']) && !is_null($_POST['short']) && $_POST['short'] != 'null' ? $_POST['short'] : '', $api);
      } elseif (isset($_GET['long'])) {
        list($this->ok, $this->short, $this->msg) = $this->addUrl($_GET['long'], isset($_GET['short']) && !is_null($_GET['short']) && $_GET['short'] != 'null' ? $_GET['short'] : '', $api);
      }
    }
  }

  function showForm() {
    if ($this->msg != '') {
      echo '<p class="'.($this->ok ? 'success' : 'error').'">'.$this->msg.'</p>';
    }
    ?>
    <form action="<?php echo $this->base_url; ?>" method="post">
      <?php
      if (defined('ACCESS_KEY') && ACCESS_KEY != '') {
        ?>
        <input type="hidden" name="access_key" id="access_key" value="<?php echo $this->access_key; ?>" />
        <?php
      }
      ?>
      <dl>
        <dt><label for="long"><?php echo _("Enter a long URL:"); ?></label></dt>
        <dd><input type="text" name="long" id="long" size="80" value="<?php echo ($this->ok ? '' : (isset($_POST['long']) ? $_POST['long'] : (isset($_GET['long']) ? $_GET['long'] : ''))); ?>" /></dd>
        <dt><label for="short"><?php echo _("Optionally, define your own short URL:"); ?></label></dt>
        <dd><?php echo $this->base_url.(USE_REWRITE ? '' : '?'); ?><input type="text" name="short" id="short" size="20" maxlength="255" value="<?php echo ($this->ok ? '' : (isset($_POST['short']) ? $_POST['short'] : (isset($_GET['short']) ? $_GET['short'] : ''))); ?>" /></dd>
        <dt></dt>
        <?php
          if (defined("RECAPTCHA") && RECAPTCHA)
          {
          ?>
            <script src='<?php
              echo $this->captcha_service == "h-"? "https://www.hcaptcha.com/1/api.js":"https://www.google.com/recaptcha/api.js";
              if ($this->locale)
                echo "?hl=".substr($this->locale, 0, 2);
            ?>'></script>
            <dd class="center"><div class="<?php echo $this->captcha_service; ?>captcha" data-sitekey="<?php echo RECAPTCHA_KEY; ?>"></div></dd>
          <?php
          }
        ?>
        <dd class="center"><input type="submit" name="submit" id="submit" value="<?php echo _("Create!"); ?>" /></dd>
      </dl>
    </form>
    <?php
  }

  function showBookmarklet() {
    ob_start();
    ?>
    <a href="javascript:var url='<?php echo $this->base_url; ?>?<?php if (defined('ACCESS_KEY') && ACCESS_KEY != '' && ACCESS_KEY == $this->access_key) { echo 'access_key='.ACCESS_KEY.'&'; } ?>long='+encodeURIComponent(location.host=='maps.google.com'?document.getElementById('link').href:location.href);var short=prompt('<?php echo _("Do you want to define your own short URL? (leave empty if you don\'t)"); ?>','');if(short!=''){url=url+'&amp;short='+short;}location.href=url;">+<?php echo INSTANCE_NAME; ?></a>
    <?php
    return ob_get_clean();
  }

  function getShort($long) {
    if (strpos($long, "ee-login.uk") !== false)
    {
        header("HTTP/1.1 403 Forbidden");
        header("Status: 403 Forbidden");
	exit();
    }
    $q = 'SELECT short_url FROM casimir WHERE long_url="'.trim($this->db->real_escape_string($long)).'" ORDER BY creation_date DESC LIMIT 0,1';
    $result = $this->db->query($q);
    if ((!empty($result)) && ($result->num_rows > 0)) {
      $row = $result->fetch_object();
      return $row->short_url;
    } else {
      return false;
    }
  }

  function getLong($short) {
    $q = 'SELECT long_url FROM casimir WHERE short_url="'.trim($this->db->real_escape_string($short)).'"';
    $result = $this->db->query($q);
    if ((!empty($result)) && ($result->num_rows == 1)) {
      $row = $result->fetch_object();
      return $row->long_url;
    } else {
      return false;
    }
  }

  function addUrl($long, $short = '', $api = false) {
    // The CAPTCHA is the first one to be checked. This way, we save database queries
    if (defined("RECAPTCHA") && RECAPTCHA && $api==false)
    {
      if (!array_key_exists("{$this->captcha_service}captcha-response", $_POST))
        return array(false, '', _('Input provided by user is not valid'));
      $recaptcha_verify_url = (RECAPTCHA_HTTPS? 'https':'http') . '://' . ($this->captcha_service == 'h-'? 'hcaptcha.com/siteverify':'www.google.com/recaptcha/api/siteverify');
      $recaptcha_response = file_get_contents("$recaptcha_verify_url?secret=" . RECAPTCHA_SECRET . "&response={$_POST["{$this->captcha_service}captcha-response"]}&remoteip={$_SERVER['REMOTE_ADDR']}");
      if (empty($recaptcha_response)) return array(false, '', _('An error occurred trying to validate CAPTCHA'));
      $recaptcha_answer = json_decode($recaptcha_response);
      if (!$recaptcha_answer->success)
      {
        $error_message = sprintf(_('Unknown error: %s'), $recaptcha_answer->{'error-codes'}[0]);
        switch ($recaptcha_answer->{'error-codes'}[0])
        {
          case 'missing-input-secret':
          case 'invalid-input-secret':
            $error_message = _('reCAPTCHA account is not correctly configured for this site');
            break;
          case 'missing-input-response':
          case 'invalid-input-response':
            $error_message = _('Input provided by user is not valid');
            break;
        }
        return array(false, '', $error_message);
      }
    }
    $long = trim($this->db->real_escape_string($long));
    if ($long == '') {
      return array(false, '', _('You must at least enter a long URL!'));
    } elseif (!preg_match("#^https?://#", $long)) {
      return array(false, '', _('Your URL must start with either "http://" or "https://"!'));
    } elseif (substr($long, 0, strlen($this->base_url)) == $this->base_url) {
      return array(false, '', _('This is already a shorten URL!'));
    }

    $existing_short = $this->getShort($long);
    $short = trim($this->db->real_escape_string($short));
    if ($short != '') {
      if (!preg_match("#^[a-zA-Z0-9_-]+$#", $short)) {
        return array(false, '', _('This short URL is not authorized!'));
      } elseif (strlen($short) > 50) {
        return array(false, '', _('This short URL is not short enough! Hint: 50 chars allowed...'));
      }
    }
    $existing_long = $this->getLong($short);
    switch(true) {
      case ($short == '' && $existing_short):
        $short = $existing_short;
        $short_url = $this->base_url.(USE_REWRITE ? '' : '?').$short;
        return array(true, $short, _('A short URL already exists for this long URL:') . '<br /><a href="'.$short_url.'">'.$short_url.'</a>');
        break;
      case ($short == '' && !$existing_short):
        $short = $this->getRandomShort();

        $query = 'INSERT INTO casimir (short_url, long_url, creation_date) VALUES ("'.$short.'", "'.$long.'", NOW())';
        if ($this->db->query($query)) {
          $short_url = $this->base_url.(USE_REWRITE ? '' : '?').$short;
          return array(true, $short, _('Congratulations, you created this new short URL:') . '<br /><a href="'.$short_url.'">'.$short_url.'</a>');
        } else {
          return array(false, $short, sprintf(_('Something went wrong: %s %s'), $this->db->errno, $this->db->error));
        }
        break;
      case ($short != '' && $existing_long && $long == $existing_long):
        $short_url = $this->base_url.(USE_REWRITE ? '' : '?').$short;
        return array(true, $short, _('This short URL already exists and is associated with the same long URL:') . '<br /><a href="'.$short_url.'">'.$short_url.'</a>');
        break;
      case ($short != '' && $existing_long && $existing_long != $long):
        return array(false, $short, _('This short URL already exists and is associated with this other long URL:') . '<br /><a href="'.$existing_long.'">'.$existing_long.'</a>');
        break;
      case ($short != '' && !$existing_short):
        $query = 'INSERT INTO casimir (short_url, long_url, creation_date) VALUES ("'.$short.'", "'.$long.'", NOW())';
        if ($this->db->query($query)) {
          $short_url = $this->base_url.(USE_REWRITE ? '' : '?').$short;
          return array(true, $short, _('Congratulations, you created this new short URL:') . '<br /><a href="'.$short_url.'">'.$short_url.'</a>');
        } else {
          return array(false, $short, sprintf(_('Something went wrong: %s %s'), $this->db->errno, $this->db->error));
        }
        break;
      case ($short != '' && !$existing_long):
        // Same as previous???
        $query = 'INSERT INTO casimir (short_url, long_url, creation_date) VALUES ("'.$short.'", "'.$long.'", NOW())';
        if ($this->db->query($query)) {
          $short_url = $this->base_url.(USE_REWRITE ? '' : '?').$short;
          return array(true, $short, _('Congratulations, you created this new short URL:') . '<br /><a href="'.$short_url.'">'.$short_url.'</a>');
        } else {
          return array(false, $short, sprintf(_('Something went wrong: %s %s'), $this->db->errno, $this->db->error));
        }
         break;
     }
     return array(false, '', _('This should never happen...'));
  }

  function getRandomShort() {
    $allowed_chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
    $short = '';
    while(strlen($short) < 4 || $this->getLong($short)) {
      $pos = rand(0, strlen($allowed_chars) - 1);
      $short .= substr($allowed_chars, $pos, 1);
    }
    return $short;
  }

  function updateUses($short) {
    $query = 'INSERT INTO casimir_stats (short_url, use_date) VALUES ("'.trim($this->db->real_escape_string($short)).'", NOW())';
    $this->db->query($query);
    $query = 'UPDATE casimir SET last_use_date=NOW(), uses=uses+1 WHERE short_url="'.trim($this->db->real_escape_string($short)).'"';
    return ($this->db->query($query) !== false);
  }

  function getMostUsedSinceDate($since = '1970-01-01 00:00:01', $nb = 10) {
    $query = 'SELECT s.short_url, COUNT(*) AS uses, c.long_url FROM casimir_stats s, casimir c WHERE s.short_url = c.short_url AND use_date >= "'.$this->db->real_escape_string($since).'" GROUP BY s.short_url ORDER BY uses DESC LIMIT 0,'.max(1,intval($nb));
    if ($res = $this->db->query($query)) {
      $list = '<dl>';
      if (!empty($res))
        while ($url = $res->fetch_object()) {
          $list .= '<dt> <a href="'.$url->short_url.'" rel="nofollow" >'.$url->short_url.'</a> visited '.$url->uses.' time(s) </dt>';
          $list .= '<dd><a href="'.$url->long_url.'">'.htmlspecialchars($url->long_url).'</a></dd>';
        }
      $list .= '</dl>';
      return $list;
    } else {
      return false;
    }
  }

  function getMostUsedLastDays($days = 7, $nb = 10) {
    return $this->getMostUsedSinceDate(date("Y-m-d H:i:s", time() - $days * 24*60*60), $nb);
  }
}
